safir/src/safir/metrics/_testing.py

```diff
@@ -0,0 +1,235 @@
+"""Helpers working with metrics events in unit tests."""
+
+from abc import ABC, abstractmethod
+from pprint import pformat
+from typing import Any, Generic, TypeVar
+from unittest.mock import ANY as MOCK_ANY
+
+from pydantic import BaseModel
+
+__all__ = [
+    "ANY",
+    "NOT_NONE",
+    "BaseAssertionError",
+    "NotPublishedConsecutivelyError",
+    "NotPublishedError",
+    "PublishedCountError",
+    "PublishedList",
+    "PublishedTooFewError",
+]
+
+P = TypeVar("P", bound=BaseModel)
+"""Generic event payload type."""
+
+
+ANY = MOCK_ANY
+"""An object that compares equal to anything, reexported from unittest.mock."""
+
+
+class _NotNone:
+    """A helper object that compares equal to everything except None."""
+
+    def __eq__(self, other: object) -> bool:
+        return other is not None
+
+    def __repr__(self) -> str:
+        return "<NOT NONE>"
+
+
+NOT_NONE = _NotNone()
+"""An object to indicate that a value can be anything except None."""
+
+
+ModelDumpList = list[dict[str, Any]]
+"""Type alias for a list of dumped Pydantic models."""
+
+
+class BaseAssertionError(Generic[P], ABC, AssertionError):
+    """Base assertion error with common attributes and messaging."""
+
+    def __init__(
+        self,
+        expected: ModelDumpList,
+        actual: ModelDumpList,
+        actual_models: list[P],
+    ) -> None:
+        self.expected = expected
+        self.actual = actual
+        self.actual_models = actual_models
+        message = (
+            f"{self.errormsg()}\n"
+            f"Expected:\n"
+            f"{pformat(self.expected)}\n"
+            f"Actual:\n"
+            f"{pformat(self.actual)}\n"
+            f"Actual models: {pformat(self.actual_models)}"
+        )
+        super().__init__(message)
+
+    @abstractmethod
+    def errormsg(self) -> str:
+        """Return a string to be added to the exception message."""
+
+
+class PublishedTooFewError(BaseAssertionError):
+    """Expected more events than have actually been published."""
+
+    def errormsg(self) -> str:
+        return "Expected more events than have actually been published"
+
+
+class PublishedCountError(BaseAssertionError):
+    """Expected has a different number of items than were published."""
+
+    def errormsg(self) -> str:
+        return "Expected has a different number of items than were published"
+
+
+class NotPublishedConsecutivelyError(BaseAssertionError):
+    """Expected events were not published consecutively."""
+
+    def errormsg(self) -> str:
+        return "Expected events were not published consecutively"
+
+
+class NotPublishedError(BaseAssertionError):
+    """Some expected items were not published."""
+
+    def __init__(
+        self,
+        expected: ModelDumpList,
+        actual: ModelDumpList,
+        actual_models: list[P],
+        not_found: ModelDumpList,
+    ) -> None:
```

Comment:
Nit: Over three parameters, so generally we would force these to be keyword parameters by adding `*,` after `self,`.